### ISSUE TITLE
Error while running on sylius

### ISSUE THREAD

**Issue by binarious** [composer-unused/composer-unused#61]
## Describe the bug
Running composer unused on [sylius](https://github.com/Sylius/Sylius) results in the following error:
```
Fatal error: Interface 'PhpParser\ErrorHandler' not found in /Users/myuser/.composer/vendor/icanhazstring/composer-unused/src/Error/ErrorHandlerInterface.php on line 10
```

Steps to reproduce:
- `composer create-project sylius/sylius-standard acme`
- `cd acme`
- `composer unused`

## Error dump
```
root@612131282c98:/acme# composer unused -vvv
Reading ./composer.json
Loading config file ./composer.json
Checked CA file /etc/ssl/certs/ca-certificates.crt: valid
Executing command (/acme): git branch --no-color --no-abbrev -v
Executing command (/acme): git describe --exact-match --tags
Executing command (/acme): git log --pretty="%H" -n1 HEAD
Executing command (/acme): hg branch
Executing command (/acme): fossil branch list
Executing command (/acme): fossil tag list
Executing command (/acme): svn info --xml
Reading /root/.composer/composer.json
Loading config file /root/.composer/composer.json
Reading /acme/vendor/composer/installed.json
Reading /root/.composer/vendor/composer/installed.json
Loading plugin Symfony\Thanks\Thanks
Loading plugin Symfony\Flex\Flex
Loading plugin PHPStan\ExtensionInstaller\Plugin
Loading plugin PackageVersions\Installer
Loading plugin Icanhazstring\Composer\Unused\UnusedPlugin
Loading plugin Hirak\Prestissimo\Plugin
PHP Fatal error:  Interface 'PhpParser\ErrorHandler' not found in /root/.composer/vendor/icanhazstring/composer-unused/src/Error/ErrorHandlerInterface.php on line 10

Fatal error: Interface 'PhpParser\ErrorHandler' not found in /root/.composer/vendor/icanhazstring/composer-unused/src/Error/ErrorHandlerInterface.php on line 10
```

## Additional information
- php 7.3

**Comment by icanhazstring** [comment]
Hi @binarious thanks for reaching out. I will look into this.

**Comment by icanhazstring** [comment]
Can you give me a list of your installed composer dependencies?

`composer show`

**Comment by binarious** [comment]
@icanhazstring Thanks for your time again.

```
root@95a0c14e3425:/acme# composer show
behat/behat                                     v3.5.0  Scenario-oriented BDD framework for PHP 5.3
behat/gherkin                                   v4.6.0  Gherkin DSL parser for PHP 5.3
behat/mink                                      v1.7.1  Browser controller/emulator abstraction for PHP
behat/mink-browserkit-driver                    1.3.3   Symfony2 BrowserKit driver for Mink framework
behat/mink-extension                            2.3.1   Mink extension for Behat
behat/mink-selenium2-driver                     v1.3.1  Selenium2 (WebDriver) driver for Mink framework
behat/transliterator                            v1.2.0  String transliterator
clue/stream-filter                              v1.4.1  A simple and modern approach to stream filtering in PHP
cocur/slugify                                   v3.2    Converts a string into a slug.
coduo/php-matcher                               3.2.2   PHP Matcher enables you to match values with patterns
coduo/php-to-string                             2.0.2
composer/ca-bundle                              1.2.4   Lets you find a path to the system CA bundle, and include...
composer/semver                                 1.5.0   Semver library that offers utilities, version constraint ...
composer/xdebug-handler                         1.3.3   Restarts a process without xdebug.
container-interop/container-interop             1.2.0   Promoting the interoperability of container objects (DIC,...
doctrine/annotations                            v1.8.0  Docblock Annotations Parser
doctrine/cache                                  v1.8.0  Caching library offering an object-oriented API for many ...
doctrine/collections                            v1.6.2  PHP Doctrine Collections library that adds additional fun...
doctrine/common                                 v2.11.0 PHP Doctrine Common project is a library that provides ad...
doctrine/data-fixtures                          v1.3.2  Data Fixtures for all Doctrine Object Managers
doctrine/dbal                                   v2.9.2  Powerful PHP database abstraction layer (DBAL) with many ...
doctrine/doctrine-bundle                        1.11.2  Symfony DoctrineBundle
doctrine/doctrine-cache-bundle                  1.3.5   Symfony Bundle for Doctrine Cache
doctrine/doctrine-fixtures-bundle               v2.4.1  Symfony DoctrineFixturesBundle
doctrine/doctrine-migrations-bundle             v2.0.0  Symfony DoctrineMigrationsBundle
doctrine/event-manager                          v1.0.0  Doctrine Event Manager component
doctrine/inflector                              v1.3.0  Common String Manipulations with regard to casing and sin...
doctrine/instantiator                           1.2.0   A small, lightweight utility to instantiate objects in PH...
doctrine/lexer                                  1.1.0   PHP Doctrine Lexer parser library that can be used in Top...
doctrine/migrations                             2.1.1   PHP Doctrine Migrations project offer additional function...
doctrine/orm                                    v2.6.4  Object-Relational-Mapper for PHP
doctrine/persistence                            1.1.1   The Doctrine Persistence project is a set of shared inter...
doctrine/reflection                             v1.0.0  Doctrine Reflection component
egulias/email-validator                         2.1.11  A library for validating emails against several RFCs
friends-of-behat/page-object-extension          v0.3.0  Provides default classes for Page object pattern in Behat
friends-of-behat/suite-settings-extension       v1.0.1  Allows to overwrite suites' default settings.
friends-of-behat/symfony-extension              v2.0.8  Integrates Behat with Symfony.
friends-of-behat/variadic-extension             v1.2.0  Variadic support for behat context arguments
friendsofphp/php-cs-fixer                       v2.15.3 A tool to automatically fix PHP code style
friendsofsymfony/oauth-server-bundle            1.6.2   Symfony2 OAuth Server Bundle
friendsofsymfony/oauth2-php                     1.2.3   OAuth2 library
friendsofsymfony/rest-bundle                    2.5.0   This Bundle provides various tools to rapidly develop RES...
fzaninotto/faker                                v1.8.0  Faker is a PHP library that generates fake data for you.
gedmo/doctrine-extensions                       v2.4.37 Doctrine2 behavioral extensions
guzzlehttp/guzzle                               6.3.3   Guzzle is a PHP HTTP client library
guzzlehttp/promises                             v1.3.1  Guzzle promises library
guzzlehttp/psr7                                 1.6.1   PSR-7 message implementation that also provides common ut...
hamcrest/hamcrest-php                           v2.0.0  This is the PHP port of Hamcrest Matchers
imagine/imagine                                 1.2.2   Image processing for PHP 5.3
instaclick/php-webdriver                        1.4.6   PHP WebDriver for Selenium 2
jdorn/sql-formatter                             v1.2.17 a PHP SQL highlighting library
jean85/pretty-package-versions                  1.2     A wrapper for ocramius/package-versions to get pretty ver...
jms/metadata                                    1.7.0   Class/method/property metadata management in PHP
jms/parser-lib                                  1.0.0   A library for easily creating recursive-descent parsers.
jms/serializer                                  1.14.0  Library for (de-)serializing data of any complexity; supp...
jms/serializer-bundle                           2.4.4   Allows you to easily serialize, and deserialize data of a...
knplabs/gaufrette                               v0.8.3  PHP library that provides a filesystem abstraction layer
knplabs/knp-gaufrette-bundle                    v0.5.3  Allows to easily use the Gaufrette library in a Symfony p...
knplabs/knp-menu                                2.6.0   An object oriented menu library
knplabs/knp-menu-bundle                         v2.3.0  This bundle provides an integration of the KnpMenu library
lakion/mink-debug-extension                     v1.2.3  Debug extension for Behat
lchrusciel/api-test-case                        v3.1.4  Perfect PHPUnit TestCase for JSON/XML API TDD with Symfony.
league/uri                                      5.3.0   URI manipulation library
league/uri-components                           1.8.2   URI components manipulation library
league/uri-hostname-parser                      1.1.1   ICANN base hostname parsing implemented in PHP.
league/uri-interfaces                           1.1.1   Common interface for URI representation
league/uri-manipulations                        1.5.0   URI manipulation library
league/uri-parser                               1.4.1   userland URI parser RFC 3986 compliant
league/uri-schemes                              1.2.1   URI manipulation library
liip/imagine-bundle                             2.2.0   This bundle provides an image manipulation abstraction to...
mockery/mockery                                 1.2.4   Mockery is a simple yet flexible PHP mock object framework
monolog/monolog                                 1.25.1  Sends your logs to files, sockets, inboxes, databases and...
myclabs/deep-copy                               1.9.3   Create deep copies (clones) of your objects
nelmio/alice                                    v3.4.1  Expressive fixtures generator
nette/finder                                    v2.5.1  🔍 Nette Finder: find files and directories with an int...
nette/robot-loader                              v3.1.1  ? Nette RobotLoader: high performance and comfortable aut...
nette/utils                                     v2.5.3  ? Nette Utils: lightweight utilities for string & array m...
ocramius/package-versions                       1.4.0   Composer plugin that provides efficient querying for inst...
ocramius/proxy-manager                          2.2.3   A library providing utilities to generate, instantiate an...
openlss/lib-array2xml                           1.0.0   Array2XML conversion library credit to lalit.org
pagerfanta/pagerfanta                           v2.1.3  Pagination for PHP
paragonie/random_compat                         v2.0.18 PHP 5.x polyfill for random_bytes() and random_int() from...
payum/iso4217                                   1.0.1   ISO 4217 PHP Library
payum/payum                                     1.5.1   One million downloads of Payum already! Payum offers ever...
payum/payum-bundle                              2.3.3   One million downloads of Payum already! Payum offers ever...
phar-io/manifest                                1.0.1   Component for reading phar.io manifest information from a...
phar-io/version                                 1.0.1   Library for handling version information and constraints
php-cs-fixer/diff                               v1.3.0  sebastian/diff v2 backport support for PHP5.6
php-http/guzzle6-adapter                        v1.1.1  Guzzle 6 HTTP Adapter
php-http/httplug                                v1.1.0  HTTPlug, the HTTP client abstraction for PHP
php-http/message                                1.8.0   HTTP Message related tools
php-http/message-factory                        v1.0.2  Factory interfaces for PSR-7 HTTP Message
php-http/promise                                v1.0.0  Promise used for asynchronous HTTP requests
phpcollection/phpcollection                     0.5.0   General-Purpose Collection Library for PHP
phpdocumentor/reflection-common                 2.0.0   Common reflection classes used by phpdocumentor to reflec...
phpdocumentor/reflection-docblock               4.3.2   With this component, a library can provide support for an...
phpdocumentor/type-resolver                     1.0.1   A PSR-5 based resolver of Class names, Types and Structur...
phpoption/phpoption                             1.5.0   Option Type for PHP
phpspec/php-diff                                v1.1.0  A comprehensive library for generating differences betwee...
phpspec/phpspec                                 5.1.2   Specification-oriented BDD framework for PHP 7.1+
phpspec/prophecy                                1.9.0   Highly opinionated mocking framework for PHP 5.3+
phpstan/extension-installer                     1.0.2   Composer plugin for automatic installation of PHPStan ext...
phpstan/phpstan-doctrine                        0.11.6  Doctrine extensions for PHPStan
phpstan/phpstan-shim                            0.11.16 PHPStan Phar distribution
phpstan/phpstan-webmozart-assert                0.11.3  PHPStan webmozart/assert extension
phpunit/php-code-coverage                       5.3.2   Library that provides collection, processing, and renderi...
phpunit/php-file-iterator                       1.4.5   FilterIterator implementation that filters files based on...
phpunit/php-text-template                       1.2.1   Simple template engine.
phpunit/php-timer                               1.0.9   Utility class for timing
phpunit/php-token-stream                        2.0.2   Wrapper around PHP's tokenizer extension.
phpunit/phpunit                                 6.5.14  The PHP Unit Testing framework.
phpunit/phpunit-mock-objects                    5.0.10  Mock Object library for PHPUnit
polishsymfonycommunity/symfony-mocker-container v1.0.2  Provides base Symfony dependency injection container enab...
psr/cache                                       1.0.1   Common interface for caching libraries
psr/container                                   1.0.0   Common Container Interface (PHP FIG PSR-11)
psr/http-message                                1.0.1   Common interface for HTTP messages
psr/log                                         1.1.0   Common interface for logging libraries
psr/simple-cache                                1.0.1   Common interfaces for simple caching
ralouphie/getallheaders                         3.0.3   A polyfill for getallheaders.
ramsey/uuid                                     3.8.0   Formerly rhumsaa/uuid. A PHP 5.4+ library for generating ...
sebastian/code-unit-reverse-lookup              1.0.1   Looks up which function or method a line of code belongs to
sebastian/comparator                            2.1.3   Provides the functionality to compare PHP values for equa...
sebastian/diff                                  2.0.1   Diff implementation
sebastian/environment                           3.1.0   Provides functionality to handle HHVM/PHP environments
sebastian/exporter                              3.1.2   Provides the functionality to export PHP variables for vi...
sebastian/global-state                          2.0.0   Snapshotting of global state
sebastian/object-enumerator                     3.0.3   Traverses array structures and object graphs to enumerate...
sebastian/object-reflector                      1.1.1   Allows reflection of object attributes, including inherit...
sebastian/recursion-context                     3.0.0   Provides functionality to recursively process PHP variables
sebastian/resource-operations                   1.0.0   Provides a list of PHP built-in functions that operate on...
sebastian/version                               2.0.1   Library that helps with managing the version number of Gi...
sensiolabs/security-checker                     v5.0.3  A security checker for your composer.lock
slam/php-cs-fixer-extensions                    v1.19.1 Slam extension of friendsofphp/php-cs-fixer
slevomat/coding-standard                        4.8.7   Slevomat Coding Standard for PHP_CodeSniffer complements ...
sonata-project/block-bundle                     3.18.1  Symfony SonataBlockBundle
sonata-project/cache                            2.0.1   Cache library
sonata-project/core-bundle                      3.17.0  Symfony SonataCoreBundle
sonata-project/datagrid-bundle                  2.5.0   Symfony SonataDatagridBundle
sonata-project/doctrine-extensions              1.3.1   Doctrine2 behavioral extensions
sonata-project/intl-bundle                      2.7.0   Symfony SonataIntlBundle
squizlabs/php_codesniffer                       3.5.0   PHP_CodeSniffer tokenizes PHP, JavaScript and CSS files a...
stof/doctrine-extensions-bundle                 v1.3.0  Integration of the gedmo/doctrine-extensions with Symfony2
stripe/stripe-php                               v6.43.1 Stripe PHP Library
swiftmailer/swiftmailer                         v6.2.1  Swiftmailer, free feature-rich PHP mailer
sylius-labs/association-hydrator                v1.1.1  Doctrine ORM hydration performance optimization made easier.
sylius-labs/coding-standard                     v2.0.0  Battle-tested coding standard configuration used in Sylius.
sylius/fixtures-bundle                          v1.4.1  Configurable fixtures for Symfony applications.
sylius/grid-bundle                              v1.6.1  Amazing grids with support of filters and custom fields i...
sylius/mailer-bundle                            v1.4.4  Mailers and e-mail template management for Symfony projects.
sylius/registry                                 v1.4.1  Services registry.
sylius/resource-bundle                          v1.6.0  Resource component for Sylius.
sylius/sylius                                   v1.6.1  E-Commerce platform for PHP, based on Symfony framework.
sylius/theme-bundle                             v1.4.6  Themes management for Symfony projects.
symfony/asset                                   v4.3.5  Symfony Asset Component
symfony/browser-kit                             v4.3.5  Symfony BrowserKit Component
symfony/cache                                   v4.3.5  Symfony Cache component with PSR-6, PSR-16, and tags
symfony/cache-contracts                         v1.1.7  Generic abstractions related to caching
symfony/class-loader                            v3.4.32 Symfony ClassLoader Component
symfony/config                                  v4.3.5  Symfony Config Component
symfony/console                                 v4.3.5  Symfony Console Component
symfony/contracts                               v1.1.0  A set of abstractions extracted out of the Symfony compon...
symfony/css-selector                            v3.4.32 Symfony CssSelector Component
symfony/debug                                   v4.3.5  Symfony Debug Component
symfony/debug-bundle                            v4.3.5  Symfony DebugBundle
symfony/dependency-injection                    v4.3.5  Symfony DependencyInjection Component
symfony/doctrine-bridge                         v4.3.5  Symfony Doctrine Bridge
symfony/dom-crawler                             v4.3.5  Symfony DomCrawler Component
symfony/dotenv                                  v4.3.5  Registers environment variables from a .env file
symfony/event-dispatcher                        v4.3.5  Symfony EventDispatcher Component
symfony/event-dispatcher-contracts              v1.1.7  Generic abstractions related to dispatching event
symfony/expression-language                     v4.3.5  Symfony ExpressionLanguage Component
symfony/filesystem                              v4.3.5  Symfony Filesystem Component
symfony/finder                                  v4.3.5  Symfony Finder Component
symfony/flex                                    v1.4.6  Composer plugin for Symfony
symfony/form                                    v4.3.5  Symfony Form Component
symfony/framework-bundle                        v4.3.5  Symfony FrameworkBundle
symfony/http-foundation                         v4.3.5  Symfony HttpFoundation Component
symfony/http-kernel                             v4.3.5  Symfony HttpKernel Component
symfony/inflector                               v4.3.5  Symfony Inflector Component
symfony/intl                                    v4.3.5  A PHP replacement layer for the C intl extension that inc...
symfony/mime                                    v4.3.5  A library to manipulate MIME messages
symfony/monolog-bridge                          v4.3.5  Symfony Monolog Bridge
symfony/monolog-bundle                          v3.4.0  Symfony MonologBundle
symfony/options-resolver                        v4.3.5  Symfony OptionsResolver Component
symfony/polyfill-ctype                          v1.12.0 Symfony polyfill for ctype functions
symfony/polyfill-iconv                          v1.12.0 Symfony polyfill for the Iconv extension
symfony/polyfill-intl-icu                       v1.12.0 Symfony polyfill for intl's ICU-related data and classes
symfony/polyfill-intl-idn                       v1.12.0 Symfony polyfill for intl's idn_to_ascii and idn_to_utf8 ...
symfony/polyfill-mbstring                       v1.12.0 Symfony polyfill for the Mbstring extension
symfony/polyfill-php70                          v1.12.0 Symfony polyfill backporting some PHP 7.0+ features to lo...
symfony/polyfill-php72                          v1.12.0 Symfony polyfill backporting some PHP 7.2+ features to lo...
symfony/polyfill-php73                          v1.12.0 Symfony polyfill backporting some PHP 7.3+ features to lo...
symfony/process                                 v4.3.5  Symfony Process Component
symfony/property-access                         v4.3.5  Symfony PropertyAccess Component
symfony/proxy-manager-bridge                    v4.3.5  Symfony ProxyManager Bridge
symfony/routing                                 v4.3.5  Symfony Routing Component
symfony/security                                v4.3.5  Symfony Security Component
symfony/security-bundle                         v4.3.5  Symfony SecurityBundle
symfony/service-contracts                       v1.1.7  Generic abstractions related to writing services
symfony/stopwatch                               v4.3.5  Symfony Stopwatch Component
symfony/swiftmailer-bundle                      v3.2.8  Symfony SwiftmailerBundle
symfony/templating                              v4.3.5  Symfony Templating Component
symfony/thanks                                  v1.1.0  Give thanks (in the form of a GitHub ⭐) to your fellow ...
symfony/translation                             v4.3.5  Symfony Translation Component
symfony/translation-contracts                   v1.1.7  Generic abstractions related to translation
symfony/twig-bridge                             v4.3.5  Symfony Twig Bridge
symfony/twig-bundle                             v4.3.5  Symfony TwigBundle
symfony/validator                               v4.3.5  Symfony Validator Component
symfony/var-dumper                              v4.3.5  Symfony mechanism for exploring and dumping PHP variables
symfony/var-exporter                            v4.3.5  A blend of var_export() + serialize() to turn any seriali...
symfony/web-profiler-bundle                     v4.3.5  Symfony WebProfilerBundle
symfony/yaml                                    v4.3.5  Symfony Yaml Component
symplify/better-phpdoc-parser                   v4.8.0  Slim wrapper around phpstan/phpdoc-parser with format pre...
symplify/coding-standard                        v4.8.0  Set of Symplify rules for PHP_CodeSniffer and PHP CS Fixer.
symplify/easy-coding-standard                   v4.8.0  Use Coding Standard with 0-knowledge of PHP-CS-Fixer and ...
symplify/package-builder                        v4.8.0  Dependency Injection, Console and Kernel toolkit for Symp...
symplify/token-runner                           v4.8.0  Set of utils for PHP_CodeSniffer and PHP CS Fixer.
thecodingmachine/safe                           v0.1.16 PHP core functions that throw exceptions instead of retur...
theofidry/alice-data-fixtures                   v1.1.1  Nelmio alice extension to persist the loaded fixtures.
theseer/tokenizer                               1.1.3   A small library for converting tokenized PHP source code ...
twig/extensions                                 v1.5.4  Common additional features for Twig that do not directly ...
twig/twig                                       v2.12.0 Twig, the flexible, fast, and secure template language fo...
webmozart/assert                                1.5.0   Assertions to validate method input/output with nice erro...
white-october/pagerfanta-bundle                 v1.2.4  Bundle to use Pagerfanta with Symfony2
willdurand/hateoas                              2.12.0  A PHP library to support implementing representations for...
willdurand/hateoas-bundle                       1.4.0   Integration of Hateoas into Symfony2.
willdurand/jsonp-callback-validator             v1.1.0  JSONP callback validator.
willdurand/negotiation                          v2.3.1  Content Negotiation tools for PHP provided as a standalon...
winzou/state-machine                            0.3.3   A very lightweight yet powerful PHP state machine
winzou/state-machine-bundle                     v0.3.1  Bundle for the very lightweight yet powerful PHP state ma...
zendframework/zend-code                         3.4.0   Extensions to the PHP Reflection API, static code scannin...
zendframework/zend-eventmanager                 3.2.1   Trigger and listen to events within a PHP application
zendframework/zend-hydrator                     2.4.2   Serialize objects to arrays, and vice versa
zendframework/zend-stdlib                       3.2.1   SPL extensions, array utilities, error handlers, and more
```

**Comment by icanhazstring** [comment]
@binarious could you do the same for global requirements please?
`composer global show`

**Comment by binarious** [comment]
@icanhazstring Sure:

```
root@95a0c14e3425:/acme# composer global show
Changed current directory to /root/.composer
hirak/prestissimo 0.3.9  composer parallel install plugin
nikic/php-parser  v4.3.0 A PHP parser written in PHP
psr/container     1.0.0  Common Container Interface (PHP FIG PSR-11)
psr/log           1.1.2  Common interface for logging libraries
symfony/finder    v4.4.0 Symfony Finder Component
```

**Comment by icanhazstring** [comment]
Thanks that helps. Looking into to 👍

**Comment by icanhazstring** [comment]
Ok I located the problem or rather the problems.

1. From what I see is `composer` does *NOT* use the global installed dependencies if they are `replaced` by another package

2. `phpstan/phpstan-shim` replaces `nikic/php-parser` which I require. So (as mentioned in 1.) the Interface can't be located because `phpstan/phpstan-shim` does not define it.

The current `dev-master` of `phpstan/phpstan-shim` has a fix, but is not released yet.
So the only solution would be to conflict with `phpstan/phpstan-shim` in its current version `0.11.19`. But then again, this does not solve the recommended installation of `composer-unused` as a global depdency.

**Comment by icanhazstring** [comment]
Only solution to this will be supplying a phar file to bundle my dependencies. So this should happen in the next couple of days.

See @62

**Comment by icanhazstring** [comment]
Phar support is now released. Unfortunately there is still a bug using it with sylius. The php-parser is running into endless recursion which breaks after a few iterations.

Problem is probably in the details. But will look into it 😉

**Comment by binarious** [comment]
For me this is fixed by updating to phpstan 0.12 and removing phpstan-shim from Sylius. Feel free to close it and thanks for your work 🙂.

**Comment by icanhazstring** [comment]
Alright. Closing it then. 👍
